File: backend-net10/Services/SupervisorService.cs
using Microsoft.EntityFrameworkCore;
using backend_net10.Data;
using backend_net10.Models;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace backend_net10.Services;

public class SupervisorService
{
    private readonly AppDbContext _context;
    private readonly ClienteService _clienteService;
    private readonly PropostaService _propostaService;

    public SupervisorService(AppDbContext context, ClienteService clienteService, PropostaService propostaService)
    {
        _context = context;
        _clienteService = clienteService;
        _propostaService = propostaService;
    }

    public async Task<List<PedidoAnalise>> ListAllAsync()
    {
        return await _context.PedidosAnalise
            .OrderByDescending(p => p.CreatedAt)
            .ToListAsync();
    }

    public async Task<List<PedidoAnalise>> ListPendentesAsync()
    {
        return await _context.PedidosAnalise
            .Where(p => p.Status == "Pendente")
            .OrderByDescending(p => p.CreatedAt)
            .ToListAsync();
    }

    public async Task<PedidoAnalise> CreateRequestAsync(PedidoAnalise request)
    {
        request.Id = Guid.NewGuid().ToString();
        request.Status = "Pendente";
        request.CreatedAt = DateTime.Now;
        request.UpdatedAt = DateTime.Now;

        _context.PedidosAnalise.Add(request);
        await _context.SaveChangesAsync();

        return request;
    }

    private class ActionPropostaReq
    {
        [JsonPropertyName("proposta_id")]
        public string PropostaId { get; set; } = string.Empty;
    }

    private class ActionEncaminharReq
    {
        [JsonPropertyName("proposta_id")]
        public string PropostaId { get; set; } = string.Empty;

        [JsonPropertyName("empresa_id")]
        public string EmpresaId { get; set; } = string.Empty;
    }

    private class ActionClienteReq
    {
        [JsonPropertyName("cliente_id")]
        public string ClienteId { get; set; } = string.Empty;
    }

    public async Task ApproveRequestAsync(string id)
    {
        var pedido = await _context.PedidosAnalise.FirstOrDefaultAsync(p => p.Id == id);
        if (pedido == null)
        {
            throw new AppNotFoundException("pedido de análise não encontrado");
        }

        if (pedido.Status != "Pendente")
        {
            throw new AppBadRequestException("este pedido de análise já foi processado");
        }

        var dadosJson = pedido.DadosAcao ?? "{}";

        switch (pedido.TipoAcao)
        {
            case "AprovarProposta":
                var aprovarReq = JsonSerializer.Deserialize<ActionPropostaReq>(dadosJson);
                if (aprovarReq == null || string.IsNullOrEmpty(aprovarReq.PropostaId))
                {
                    throw new AppBadRequestException("dados da ação inválidos para aprovar proposta");
                }

                var prop = await _propostaService.GetByIdAsync(aprovarReq.PropostaId);
                prop.Status = "Aprovada";
                await _propostaService.HandleBrokerageCommissionAsync(prop);
                await _propostaService.UpdateAsync(prop.Id, prop);
                break;

            case "EncaminharEmpresa":
                var encaminharReq = JsonSerializer.Deserialize<ActionEncaminharReq>(dadosJson);
                if (encaminharReq == null || string.IsNullOrEmpty(encaminharReq.PropostaId) || string.IsNullOrEmpty(encaminharReq.EmpresaId))
                {
                    throw new AppBadRequestException("dados da ação inválidos para encaminhar empresa");
                }

                var propEnc = await _propostaService.GetByIdAsync(encaminharReq.PropostaId);
                propEnc.EmpresaId = encaminharReq.EmpresaId;
                propEnc.StatusCorretagem = "Encaminhada";
                await _propostaService.HandleBrokerageCommissionAsync(propEnc);
                await _propostaService.UpdateAsync(propEnc.Id, propEnc);
                break;

            case "DeletarCliente":
                var deletarCliReq = JsonSerializer.Deserialize<ActionClienteReq>(dadosJson);
                if (deletarCliReq == null || string.IsNullOrEmpty(deletarCliReq.ClienteId))
                {
                    throw new AppBadRequestException("dados da ação inválidos para deletar cliente");
                }

                await _clienteService.DeleteAsync(deletarCliReq.ClienteId);
                break;

            case "DeletarProposta":
                var deletarPropReq = JsonSerializer.Deserialize<ActionPropostaReq>(dadosJson);
                if (deletarPropReq == null || string.IsNullOrEmpty(deletarPropReq.PropostaId))
                {
                    throw new AppBadRequestException("dados da ação inválidos para deletar proposta");
                }

                await _propostaService.DeleteAsync(deletarPropReq.PropostaId);
                break;

            default:
                throw new AppBadRequestException("tipo de ação do supervisor não suportado");
        }

        pedido.Status = "Aprovado";
        pedido.UpdatedAt = DateTime.Now;
        await _context.SaveChangesAsync();
    }

    public async Task RejectRequestAsync(string id)
    {
        var pedido = await _context.PedidosAnalise.FirstOrDefaultAsync(p => p.Id == id);
        if (pedido == null)
        {
            throw new AppNotFoundException("pedido de análise não encontrado");
        }

        if (pedido.Status != "Pendente")
        {
            throw new AppBadRequestException("este pedido de análise já foi processado");
        }

        pedido.Status = "Recusado";
        pedido.UpdatedAt = DateTime.Now;
        await _context.SaveChangesAsync();
    }
}

File: backend-springboot/src/main/java/br/com/portalproposta/backend/service/SupervisorService.java
package br.com.portalproposta.backend.service;

import br.com.portalproposta.backend.exception.AppBadRequestException;
import br.com.portalproposta.backend.exception.AppNotFoundException;
import br.com.portalproposta.backend.model.PedidoAnalise;
import br.com.portalproposta.backend.model.Proposta;
import br.com.portalproposta.backend.repository.PedidoAnaliseRepository;
import tools.jackson.core.JacksonException;
import tools.jackson.databind.JsonNode;
import tools.jackson.databind.ObjectMapper;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.List;
import java.util.UUID;

@Service
public class SupervisorService {

    private final PedidoAnaliseRepository pedidoAnaliseRepository;
    private final ClienteService clienteService;
    private final PropostaService propostaService;
    private final ObjectMapper objectMapper = new ObjectMapper();

    @Autowired
    public SupervisorService(PedidoAnaliseRepository pedidoAnaliseRepository,
            ClienteService clienteService,
            PropostaService propostaService) {
        this.pedidoAnaliseRepository = pedidoAnaliseRepository;
        this.clienteService = clienteService;
        this.propostaService = propostaService;
    }

    public List<PedidoAnalise> listAll() {
        return pedidoAnaliseRepository.findAllByOrderByCreatedAtDesc();
    }

    public List<PedidoAnalise> listPendentes() {
        return pedidoAnaliseRepository.findByStatusOrderByCreatedAtDesc("Pendente");
    }

    @Transactional
    public PedidoAnalise createRequest(PedidoAnalise request) {
        request.setId(UUID.randomUUID().toString());
        request.setStatus("Pendente");
        request.setCreatedAt(LocalDateTime.now());
        request.setUpdatedAt(LocalDateTime.now());

        return pedidoAnaliseRepository.save(request);
    }

    @Transactional
    public void approveRequest(String id) {
        PedidoAnalise pedido = pedidoAnaliseRepository.findById(id)
                .orElseThrow(() -> new AppNotFoundException("pedido de análise não encontrado"));

        if (!"Pendente".equals(pedido.getStatus())) {
            throw new AppBadRequestException("este pedido de análise já foi processado");
        }

        String dadosJson = pedido.getDadosAcao() != null ? pedido.getDadosAcao() : "{}";

        try {
            JsonNode root = objectMapper.readTree(dadosJson);

            switch (pedido.getTipoAcao()) {
                case "AprovarProposta": {
                    if (root.get("proposta_id") == null || root.get("proposta_id").asText().isEmpty()) {
                        throw new AppBadRequestException("dados da ação inválidos para aprovar proposta");
                    }
                    String propostaId = root.get("proposta_id").asText();
                    Proposta prop = propostaService.getById(propostaId);
                    prop.setStatus("Aprovada");
                    propostaService.handleBrokerageCommission(prop);
                    propostaService.update(prop.getId(), prop);
                    break;
                }
                case "EncaminharEmpresa": {
                    if (root.get("proposta_id") == null || root.get("proposta_id").asText().isEmpty() ||
                            root.get("empresa_id") == null || root.get("empresa_id").asText().isEmpty()) {
                        throw new AppBadRequestException("dados da ação inválidos para encaminhar empresa");
                    }
                    String propostaId = root.get("proposta_id").asText();
                    String empresaId = root.get("empresa_id").asText();

                    Proposta prop = propostaService.getById(propostaId);
                    prop.setEmpresaId(empresaId);
                    prop.setStatusCorretagem("Encaminhada");
                    propostaService.handleBrokerageCommission(prop);
                    propostaService.update(prop.getId(), prop);
                    break;
                }
                case "DeletarCliente": {
                    if (root.get("cliente_id") == null || root.get("cliente_id").asText().isEmpty()) {
                        throw new AppBadRequestException("dados da ação inválidos para deletar cliente");
                    }
                    String clienteId = root.get("cliente_id").asText();
                    clienteService.delete(clienteId);
                    break;
                }
                case "DeletarProposta": {
                    if (root.get("proposta_id") == null || root.get("proposta_id").asText().isEmpty()) {
                        throw new AppBadRequestException("dados da ação inválidos para deletar proposta");
                    }
                    String propostaId = root.get("proposta_id").asText();
                    propostaService.delete(propostaId);
                    break;
                }
                default:
                    throw new AppBadRequestException("tipo de ação do supervisor não suportado");
            }
        } catch (JacksonException e) {
            throw new AppBadRequestException("dados da ação malformados ou inválidos");
        }

        pedido.setStatus("Aprovado");
        pedido.setUpdatedAt(LocalDateTime.now());
        pedidoAnaliseRepository.save(pedido);
    }

    @Transactional
    public void rejectRequest(String id) {
        PedidoAnalise pedido = pedidoAnaliseRepository.findById(id)
                .orElseThrow(() -> new AppNotFoundException("pedido de análise não encontrado"));

        if (!"Pendente".equals(pedido.getStatus())) {
            throw new AppBadRequestException("este pedido de análise já foi processado");
        }

        pedido.setStatus("Recusado");
        pedido.setUpdatedAt(LocalDateTime.now());
        pedidoAnaliseRepository.save(pedido);
    }
}
